packages/headless/src/controllers/commerce/product-listing/sort/headless-product-listing-sort.ts

```diff
@@ -0,0 +1,159 @@
+import {CoreEngine} from '../../../../app/engine';
+import {
+  SortByFields,
+  SortCriterion,
+  sortCriterionDefinition,
+  SortByFieldsFields,
+  SortByRelevance,
+  SortBy,
+  buildRelevanceSortCriterion,
+  SortDirection,
+  buildFieldsSortCriterion,
+} from '../../../../features/sort/sort';
+import {validateInitialState} from '../../../../utils/validate-payload';
+import {
+  buildController,
+  Controller,
+} from '../../../controller/headless-controller';
+import {loadReducerError} from '../../../../utils/errors';
+import {ProductListingV2Section} from '../../../../state/state-sections';
+import {CommerceEngine} from '../../../../app/commerce-engine/commerce-engine';
+import {Schema} from '@coveo/bueno';
+import {fetchProductListing} from '../../../../features/commerce/product-listing/product-listing-actions';
+import {sortReducer as sort} from '../../../../features/commerce/product-listing/sort/product-listing-sort-slice';
+import {productListingV2Reducer as productListing} from '../../../../features/commerce/product-listing/product-listing-slice';
+import {updatePage} from '../../../../features/pagination/pagination-actions';
+import {applySort} from '../../../../features/commerce/product-listing/sort/product-listing-sort-actions';
+
+export type {SortByRelevance, SortByFields, SortByFieldsFields, SortCriterion};
+export {
+  SortBy,
+  SortDirection,
+  buildRelevanceSortCriterion,
+  buildFieldsSortCriterion,
+};
+
+export interface SortProps {
+  /**
+   * The initial state that should be applied to this `Sort` controller.
+   */
+  initialState?: SortInitialState;
+}
+
+export interface SortInitialState {
+  /**
+   * The initial sort criterion to register in state.
+   */
+  criterion?: SortCriterion;
+}
+
+function validateSortInitialState(
+  engine: CoreEngine<ProductListingV2Section>,
+  state: SortInitialState | undefined
+) {
+  if (!state) {
+    return;
+  }
+
+  const schema = new Schema<SortInitialState>({
+    criterion: sortCriterionDefinition,
+  });
+
+  validateInitialState(engine, schema, state, 'buildSort');
+}
+
+export interface Sort extends Controller {
```

Comment:
So, at a quick glance, my initial reaction for this whole file, was that "there must be some overlap/more code reuse we could do with `headless-core-sort.ts`".

But when looking into it in more depth, I'd say in general what we have here is generally good 👍 

One thing we could want to make sure is that this interface here (or the interface for `SortProps`) does not differ too much than what we have for classic search use case for no good reason.

There could be a way to use generics to make sure that for both "search API sort" and "commerce service sort" controller, we have the same `sortBy`, or `isSortedBy` method, but that what they accept as parameter are different.